crates/core/component/funding/src/component/liquidity_tournament/votes/mod.rs

```diff
@@ -2,43 +2,195 @@ use std::pin::Pin;
+                state_key::lqt::v1::votes::by_asset::ranked::parse_key(&k)
+                    .expect("should be able to parse lqt::v1::votes::by_asset::ranked key")
+            })
             .boxed()
     }
+
+    fn ranked_voters(
+        &self,
+        epoch: u64,
+        asset: asset::Id,
+    ) -> Pin<Box<dyn Stream<Item = anyhow::Result<((u64, Address), TransactionId)>> + Send + 'static>>
+    {
+        self
+        .nonverifiable_prefix_raw(&state_key::lqt::v1::votes::by_voter::ranked::prefix(
+            epoch,
+            asset
+        ))
+        .map_ok(|(k, v)| {
+            let key = state_key::lqt::v1::votes::by_voter::ranked::parse_key(&k)
+                .expect("should be able to parse lqt::v1::votes::by_voter::ranked key");
+            let value = TransactionId(v.try_into().expect("should be able to parse transaction hash from lqt::v1::votes::by_voter::ranked_key"));
+            (key, value)
+        }).boxed()
+    }
 }
 
-impl<T: StateRead + ?Sized> StateReadExt for T {}
+impl<T: StateRead> StateReadExt for T {}
+
+fn add_power(total: u64, power: u64) -> u64 {
+    total.checked_add(power).expect("LQT voting power overflow")
+}
 
 #[async_trait]
-pub trait StateWriteExt: StateWrite {
+pub trait StateWriteExt: StateWrite + Sized {
     // Keeping this as returning a result to not have to touch other code if it changes to return an error.
-    async fn tally(
-        &mut self,
-        epoch: u64,
-        asset: asset::Id,
-        power: u64,
-        voter: &Address,
-    ) -> anyhow::Result<()> {
-        self.nonverifiable_put_raw(
-            state_key::lqt::v1::votes::receipt(epoch, asset, power, voter).to_vec(),
-            Vec::default(),
-        );
-        Ok(())
+    async fn tally(&mut self, epoch: u64, asset: asset::Id, power: u64, voter: &Address) {
+        let tx = self.get_current_source().expect("source should be set");
+        // Increment total.
+        {
+            let current = get_total(&self, epoch).await;
+            put_total(&mut *self, epoch, add_power(current, power)).await;
+        }
+        // Adjust asset indices.
+        {
+            let current = get_asset_total(&self, epoch, asset).await;
+            delete_ranked_asset(&mut *self, epoch, current, asset);
+            let new = add_power(current, power);
+            put_asset_total(&mut *self, epoch, asset, new).await;
+            put_ranked_asset(&mut *self, epoch, new, asset);
+        }
+        // Adjust voter indices.
+        {
+            let current = get_voter_total(&self, epoch, asset, voter).await;
+            delete_ranked_voter(&mut *self, epoch, asset, current, voter);
+            let new = add_power(current, power);
+            put_voter_total(&mut *self, epoch, asset, voter, new).await;
+            put_ranked_voter(&mut self, epoch, asset, new, voter, tx);
+        }
     }
 }
 
-impl<T: StateWrite + ?Sized> StateWriteExt for T {}
+impl<T: StateWrite> StateWriteExt for T {}
```

Comment:
Curious why restrict the bound here? the ?Sized is to work with trait objs iirc

Reply:
I ended up needing to add the sized constraint because the guts of the method now need borrows, iirc